### ISSUE TITLE
floating point precision changes between julia versions

### ISSUE THREAD

**Issue by Kainka** [JuliaLang/julia#40525]
I've encountered a problem with floating point precision when upgrading versions. On projects with Julia support for v1.3 - 1.5.3 (inclusive) the below reproducible example generated different outputs to v1.5.4 and v1.6.0
This was run on Linux 64-bit x86 julia versions.

**in v1.3 - 1.5.3** 

`julia> 1 * 10. ^ (-5)`
`output: 1.0e-5`

**in v1.5.4 ad v1.6.0:**

`julia> 1 * 10. ^ (-5)`
`output: 9.999999999999999e-6`

I managed to workaround it by changing to this:

`julia> 1 * (10. ^ (5))^ (-1)`
`output: 1.0e-5`

which gave me the same output across all supported versions. 
It seems to be only a problem with the power of (-5) other cases are consistent across the versions, e.g.:
`julia> 1 * 10. ^ (-7)` gives `1.0e-7`

Is this an expected change from v1.5.4+?

**Comment by JeffBezanson** [comment]
Both versions call `llvm.pow.f64`. Defining `f() = 10. ^ -5`, in 1.5 I get
```
julia> @code_llvm f()

;  @ REPL[2]:1 within `f'
define double @julia_f_276() {
top:
  ret double 1.000000e-05
}
```
and in later versions I get
```
define double @julia_f_530() #0 {
top:
  ret double 0x3EE4F8B588E368F0
}

julia> reinterpret(Float64, 0x3EE4F8B588E368F0)
9.999999999999999e-6
```
so it looks like LLVM is doing this?

**Comment by vtjnash** [comment]
Possibly would be fixed by https://github.com/JuliaLang/julia/pull/40236 though?

**Comment by JeffBezanson** [comment]
Doesn't that still use the llvm intrinsic for Float64 pow though?

**Comment by oscardssmith** [comment]
Yeah. I haven't had the time to fix `Float64` pow since it requires higher precision `log2` and `exp2` than `Base` currently has (specifically, it needs `log2` that outputs to 2 `Float64` with about 70 bits of precision, and an `exp2` that takes in 2 `Float64` as input.

**Comment by chilledgeek** [comment]
Just to chip in and share my observations (i'm using julia v1.5.3), i found a way to avoid this floating point precision problem by doing direct division instead

```
julia> 8542 * 1e-4
0.8542000000000001

julia> 8542 / 1e4
0.8542
```

I have noticed that this doesn't seem to be a one off behaviour. Here's another example with increments of 0.0001 starting from 0.8000 to 0.8030 showing some interesting behaviour where the multiplying `1e-4` is different to dividing by `1e4` (but not for ALL values)

```
julia> for x in 8000:8030
           multiplication = x * 1e-4
           division = x / 1e4
           if multiplication != division
               println("$multiplication vs $division")
           end
       end
0.8006000000000001 vs 0.8006
0.8007000000000001 vs 0.8007
0.8008000000000001 vs 0.8008
0.8009000000000001 vs 0.8009
0.8016000000000001 vs 0.8016
0.8017000000000001 vs 0.8017
0.8018000000000001 vs 0.8018
0.8019000000000001 vs 0.8019
0.8026000000000001 vs 0.8026
0.8027000000000001 vs 0.8027
0.8028000000000001 vs 0.8028
0.8029000000000001 vs 0.8029
```

the problem seems to go beyond the number of decimal places...i'm seeing it happen for 5, 6, 7 decimal places as well:

```
julia> for x in 9000000:9000030
           multiplication = x * 1e-7
           division = x / 1e7
           if multiplication != division
               println("$multiplication vs $division")
           end
       end
0.8999999999999999 vs 0.9
0.9000001999999999 vs 0.9000002
0.9000003999999999 vs 0.9000004
0.9000005999999999 vs 0.9000006
0.9000007999999999 vs 0.9000008
0.9000009999999999 vs 0.900001
0.9000016999999999 vs 0.9000017
0.9000018999999999 vs 0.9000019
0.9000020999999999 vs 0.9000021
0.9000022999999999 vs 0.9000023
0.9000024999999999 vs 0.9000025
0.9000026999999999 vs 0.9000027
0.9000028999999999 vs 0.9000029
```

i haven't tested on other versions, but just switching to division seems to solve this issue for me. 

Hope this info is useful!

**Comment by martinholters** [comment]
That is unrelated and just a consequence of `1e-4` not being exactly representable, and division by 1e4 is not equal to multiplication with a value that is just close to 1e-4.

**Comment by vtjnash** [comment]
Seems fixed by c8ae5c60f52, since we appear to no longer constant fold it (@oscardssmith)
